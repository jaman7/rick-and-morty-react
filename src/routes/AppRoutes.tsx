import { Routes, Route } from 'react-router-dom';
import { motion } from 'framer-motion';
import { lazy, Suspense } from 'react';
import PageLoader from '@/shared/components/PageLoader/PageLoader';
import MainLayout from '@/view/components/layouts/MainLayout';

const MainPage = lazy(() => import('@/view/pages/MainPage/MainPage'));
const EpisodesPage = lazy(() => import('@/view/pages/EpisodesPage/EpisodesPage'));
const LocationPage = lazy(() => import('@/view/pages/LocationPage/LocationPage'));

const PageWrapper = ({ children }: { children: React.ReactNode }) => (
  <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} exit={{ opacity: 0, y: -20 }} transition={{ duration: 0.3 }}>
    {children}
  </motion.div>
);

const AppRoutes = () => {
  return (
    <Suspense fallback={<PageLoader />}>
      <Routes>
        <Route path="/" element={<MainLayout />}>
          <Route
            path="/"
            element={
              <PageWrapper>
                <MainPage />
              </PageWrapper>
            }
          />
          <Route
            path="/episodes"
            element={
              <PageWrapper>
                <EpisodesPage />
              </PageWrapper>
            }
          />
          <Route
            path="/location"
            element={
              <PageWrapper>
                <LocationPage />
              </PageWrapper>
            }
          />
        </Route>
      </Routes>
    </Suspense>
  );
};

export default AppRoutes;
