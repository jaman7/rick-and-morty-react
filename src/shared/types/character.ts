export type ICharacterStatusType = 'Alive' | 'Dead' | 'unknown';

export type ICharacterGenderType = 'Male' | 'Female' | 'Genderless' | 'unknown';

export interface ICharacterOrigin {
  name: string;
  url: string;
}

export interface ICharacterLocation {
  name: string;
  url: string;
}

export interface ICharacter {
  id: number;
  name: string;
  status: ICharacterStatusType;
  species: string;
  type: string;
  gender: ICharacterGenderType;
  origin: ICharacterOrigin;
  location: ICharacterLocation;
  image: string;
  episode: string[];
  url: string;
  created: string;
}

export interface IPaginatedResponse {
  count?: number;
  pages?: number;
  next?: string | null;
  prev?: string | null;
}

export interface ICharactersResponse {
  info: IPaginatedResponse;
  results: ICharacter[];
}

export interface IEpisode {
  id: number;
  name: string;
  air_date: string;
  episode: string;
  characters: string[];
  url: string;
  created: string;
}

export interface ILocation {
  id: number;
  name: string;
  type: string;
  dimension: string;
  residents: string[];
  url: string;
  created: string;
}

export interface IEpisodesResponse {
  info: IPaginatedResponse;
  results: IEpisode[];
}

export interface ILocationResponse {
  info: IPaginatedResponse;
  results: ILocation[];
}
