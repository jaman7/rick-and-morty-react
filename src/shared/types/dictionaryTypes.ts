export interface IDictType {
  displayName?: string;
  id?: number | string;
  [name: string]: any;
}

export interface IDictionaries {
  [name: string]: IDictType[];
}
