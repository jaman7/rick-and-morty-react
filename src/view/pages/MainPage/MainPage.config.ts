import { FilterConfig } from '@/shared/components/FiltersPanel/FiltersPanel';

const speciesDict = ['Human', 'Alien', 'Humanoid', 'Poopybutthole', 'Mythological', 'Unknown', 'Animal', 'Disease', 'Robot', 'Cronenberg'];

const gendersDict = ['Female', 'Male', 'Genderless', 'Unknown'];

const statusDict = ['Alive', 'Dead', 'Unknown'];

export const filtersConfig: FilterConfig = {
  status: { value: statusDict },
  species: { value: speciesDict },
  genders: { value: gendersDict },
};
