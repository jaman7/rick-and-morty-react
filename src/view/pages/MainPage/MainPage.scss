@use "sass:color";
@import "@scss/mixins";

.page-layout {
  display: grid;
  gap: 1rem;
  grid-template-columns: 240px 1fr;

  &.with-fixed-panel {
    grid-template-columns: 1fr;
  }

  &.without-pinned-panel {
    grid-template-columns: 32px 1fr;
  }

  @include mq(maxsm) {
    grid-template-columns: 200px 1fr;
  }

  transition: all 0.3s ease-in-out;
}
