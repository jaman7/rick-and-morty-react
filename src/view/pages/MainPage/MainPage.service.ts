import { IParams } from '@/core/services/http/http.models';
import HttpService from '@/core/services/http/http.service';
import { ICharactersResponse } from '@/shared/types/character';
import { catchError, from, map, Observable, of } from 'rxjs';

export const apiCharacters = 'https://rickandmortyapi.com/api/character';

const http = new HttpService();

export const getCharacters = (params: IParams): Observable<ICharactersResponse> => {
  return from(http.get<ICharactersResponse>(`${apiCharacters}/`, params)).pipe(
    map((response) => {
      return response ?? {};
    }),
    catchError((error) => {
      console.error('Error fetching characters', error);
      return of(error);
    })
  );
};
