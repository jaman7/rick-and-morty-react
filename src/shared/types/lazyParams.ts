import { IDictType } from '@/shared/types/dictionaryTypes'; // Replace with the correct path

export type ITableIconsType = 'VIEW' | 'EDIT' | 'DELETE' | 'ASSIGN';

export type ITableColumnsType = 'Text' | 'DateTime' | 'Checkbox' | 'Boolean' | 'UserList';

export type MatchModeTypes = 'equals' | 'startsWith' | 'contains' | 'in' | 'notContains' | 'endsWith' | 'notEquals' | 'inputIn';

export interface ITableColumns {
  field?: string;
  header?: string;
  type?: ITableColumnsType;
  sortField?: string;
  sortable?: boolean;
  formatDate?: string;
  dictName?: string;
  dictionary?: IDictType[];
  customHeader?: string;
  userLogoSize?: number;
  format?: string;
  diableFiltrSort?: boolean;
  filterType?: 'select' | 'input' | 'datepicker' | 'checkbox';

  filter?: {
    field?: string;
    type?: 'select' | 'input' | 'datepicker' | 'checkbox';
    matchMode?: MatchModeTypes;
    options?: { label: string; value: any }[];
    placeholder?: string;
    dateTimeType?: 'month' | 'year' | 'date';
    dateTimeMode?: 'single' | 'range';
    format?: string;
  };
  customizeValue?: <T>(val: T) => T | string;
}

export interface ILazyState {
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  sort?: { field?: string | null; order?: undefined | null | 0 | 1 | -1 };
  filters?: Record<string, { value: any; matchMode?: MatchModeTypes }>;
  page?: number;
  pageSize?: number;
  pages?: number;
}
