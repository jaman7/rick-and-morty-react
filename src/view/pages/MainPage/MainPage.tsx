import { useCallback, useMemo, useState } from 'react';
import { apiCharacters, getCharacters } from './MainPage.service';
import { catchError, of, tap } from 'rxjs';
import { ICharacter } from '@/shared/types/character';
import { setTableParams } from '@/shared/utils/table-utils';
import { IParams } from '@/core/services/http/http.models';
import Card from '@/shared/components/Card/Card';
import Paginator from '@/shared/components/Paginator/Paginator';
import AutoComplete from '@/shared/components/AutoComplete/AutoComplete';
import FiltersPanel from '@/shared/components/FiltersPanel/FiltersPanel';
import { filtersConfig } from './MainPage.config';
import classNames from 'classnames';
import { ILazyState } from '@/shared/types/lazyParams';
import { useToast } from '@/hooks/useToast';
import './MainPage.scss';

const MainPage: React.FC = () => {
  const [data, setData] = useState<ICharacter[]>([]);
  const [params, setParams] = useState<IParams>({ page: 1 });
  const [isFilterPinned, setIsFilterPinned] = useState(false);
  const [isFilterCollapsed, setIsFilterCollapsed] = useState(false);

  const toast = useToast();

  const containerClass = useMemo(
    () =>
      classNames('page-layout', {
        'with-fixed-panel': isFilterPinned,
        'collapsed-panel': isFilterCollapsed,
        'without-pinned-panel': !isFilterPinned && isFilterCollapsed,
      }),
    [isFilterPinned, isFilterCollapsed]
  );

  const getData = useCallback(
    (event?: ILazyState) => {
      const baseParams = { ...setTableParams(event as ILazyState) };
      if (!baseParams?.page) return;
      const fullParams = {
        ...baseParams,
      };
      setParams(fullParams);
      if (fullParams?.pageSize) delete fullParams.pageSize;
      const { pages, ...rest } = fullParams || {};
      const subscription = getCharacters(rest)
        .pipe(
          tap((res) => {
            const { results = [], info = {} } = res || {};
            if (results.length) {
              setParams((prevParams) => ({ ...prevParams, pages: info.pages }));
              setData(results ?? []);
            } else {
              toast.showError('Error fetching characters');
            }
          }),
          catchError((error) => {
            console.error('Error fetching characters', error);
            return of({});
          })
        )
        .subscribe();
      return () => subscription.unsubscribe();
    },
    [params]
  );

  return (
    <div className="main-page-container">
      <h2 className="page-title">Characters</h2>

      <AutoComplete
        url={apiCharacters}
        queryVal={(params?.name as string) ?? ''}
        paramName="name"
        onSelect={(name) => getData({ ...params, page: 1, filters: { ['name']: { value: name, matchMode: 'contains' } } })}
        className="my-3"
      />

      <div className="paginator-top">
        <Paginator
          currentPage={typeof params.page === 'number' ? params.page : 1}
          totalPages={typeof params.pages === 'number' ? params.pages : 1}
          onPageChange={(newPage) => getData({ ...params, page: newPage })}
          maxVisiblePages={3}
        />
      </div>

      <div className={containerClass}>
        <FiltersPanel
          isPinned={isFilterPinned}
          onPinChange={setIsFilterPinned}
          onCollapseChange={setIsFilterCollapsed}
          config={filtersConfig}
          onChange={(filters) => {
            getData({ ...params, page: 1, filters });
          }}
        />

        <div className="card-grid">
          {data?.map((item, i) => (
            <Card key={`card-main-${i}`} hoverable={true} name={item.name} image={item.image}>
              <div className="main-card-content">
                <div className="main-card-content__status">
                  Status: <span className={`status status--${item.status.toLowerCase()}`}>{item.status}</span>
                </div>
                <div className="main-card-content__species">
                  Species: <span>{item.species}</span>
                </div>

                <div className="card-footer">
                  <span>Origin: {item.origin.name}</span>
                </div>
              </div>
            </Card>
          ))}
        </div>
      </div>

      <div className="paginator-bottom">
        <Paginator
          currentPage={typeof params.page === 'number' ? params.page : 1}
          totalPages={typeof params.pages === 'number' ? params.pages : 1}
          onPageChange={(newPage) => getData({ ...params, page: newPage })}
        />
      </div>
    </div>
  );
};

export default MainPage;
