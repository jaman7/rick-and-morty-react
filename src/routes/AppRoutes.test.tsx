import { describe, it, expect, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import AppRoutes from './AppRoutes';
import { MemoryRouter } from 'react-router-dom';
import { of } from 'rxjs';
import React from 'react';
import * as service from '@/view/pages/MainPage/MainPage.service';
import { ICharactersResponse } from '@/shared/types/character';

vi.mock('@/shared/components/PageLoader/PageLoader', () => ({
  default: () => <div>Loading...</div>,
}));

const showErrorMock = vi.fn();

const mockCharacters: ICharactersResponse = {
  info: {
    count: 826,
    pages: 42,
    next: 'https://rickandmortyapi.com/api/character/?page=2',
    prev: null,
  },
  results: [
    {
      id: 1,
      name: 'Rick Sanchez',
      status: 'Alive' as const,
      species: 'Human',
      type: '',
      gender: 'Male',
      origin: {
        name: 'Earth (C-137)',
        url: 'https://rickandmortyapi.com/api/location/1',
      },
      location: {
        name: 'Citadel of Ricks',
        url: 'https://rickandmortyapi.com/api/location/3',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
      episode: ['https://rickandmortyapi.com/api/episode/1'],
      url: 'https://rickandmortyapi.com/api/character/1',
      created: '2017-11-04T18:48:46.250Z',
    },
    {
      id: 2,
      name: 'Morty Smith',
      status: 'Alive' as const,
      species: 'Human',
      type: '',
      gender: 'Male',
      origin: {
        name: 'unknown',
        url: '',
      },
      location: {
        name: 'Citadel of Ricks',
        url: 'https://rickandmortyapi.com/api/location/3',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
      episode: ['https://rickandmortyapi.com/api/episode/1'],
      url: 'https://rickandmortyapi.com/api/character/2',
      created: '2017-11-04T18:50:21.651Z',
    },
    {
      id: 3,
      name: 'Summer Smith',
      status: 'Alive',
      species: 'Human',
      type: '',
      gender: 'Female',
      origin: {
        name: 'Earth (Replacement Dimension)',
        url: 'https://rickandmortyapi.com/api/location/20',
      },
      location: {
        name: 'Earth (Replacement Dimension)',
        url: 'https://rickandmortyapi.com/api/location/20',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/3.jpeg',
      episode: ['https://rickandmortyapi.com/api/episode/6'],
      url: 'https://rickandmortyapi.com/api/character/3',
      created: '2017-11-04T19:09:56.428Z',
    },
  ],
};

// Helper
const renderWithRouter = (initialRoute = '/') => {
  return render(
    <MemoryRouter initialEntries={[initialRoute]}>
      <AppRoutes />
    </MemoryRouter>
  );
};

describe('AppRoutes', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.spyOn(service, 'getCharacters').mockReturnValue(of(mockCharacters));
    showErrorMock.mockReset();
  });

  it('matches snapshot', () => {
    const { asFragment } = renderWithRouter('/');
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows PageLoader while loading', () => {
    renderWithRouter('/');
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('shows PageLoader while loading lazy components', async () => {
    const LazyComponent = React.lazy(
      () => new Promise<{ default: React.FC }>((res) => setTimeout(() => res({ default: () => <h2>Lazy Loaded</h2> }), 100))
    );

    const FallbackTest = () => (
      <MemoryRouter>
        <React.Suspense fallback={<div>Loading...</div>}>
          <LazyComponent />
        </React.Suspense>
      </MemoryRouter>
    );

    render(<FallbackTest />);
    expect(screen.getByText('Loading...')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByText('Lazy Loaded')).toBeInTheDocument();
    });
  });

  it('renders MainPage at /', async () => {
    renderWithRouter('/');
    await waitFor(async () => {
      expect(screen.getByText('Rick Sanchez')).toBeInTheDocument();
      expect(screen.getByText('Morty Smith')).toBeInTheDocument();
      expect(screen.getByPlaceholderText(/Search character.../i)).toBeInTheDocument();
      await screen.findByText((content, node) => node?.tagName === 'H2' && /characters/i.test(content));
    });
  });

  it('renders EpisodesPage at /episodes', async () => {
    renderWithRouter('/episodes');
    await waitFor(() => {
      expect(screen.getByRole('heading', { name: /episodes/i })).toBeInTheDocument();
    });
  });

  it('renders LocationPage at /location', async () => {
    renderWithRouter('/location');
    await waitFor(() => {
      expect(screen.getByRole('heading', { name: /location/i })).toBeInTheDocument();
    });
  });
});
