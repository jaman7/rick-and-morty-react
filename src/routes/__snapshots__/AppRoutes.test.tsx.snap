// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppRoutes > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
